contracts/utils/VaultQueue.sol

```diff
@@ -0,0 +1,118 @@
+// SPDX-License-Identifier: MIT
+pragma solidity =0.8.4;
+
+import "@openzeppelin/contracts/access/Ownable.sol";
+import {
+    ReentrancyGuardUpgradeable
+} from "@openzeppelin/contracts-upgradeable/security/ReentrancyGuardUpgradeable.sol";
+import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import {ISAVAX} from "../interfaces/ISAVAX.sol";
+import {IRibbonVault} from "../interfaces/IRibbon.sol";
+import {IRibbonThetaVault} from "../interfaces/IRibbonThetaVault.sol";
+import {Vault} from "../libraries/Vault.sol";
+import {DSMath} from "../vendor/DSMath.sol";
+
+interface IVaultQueue {
+    struct Transfer {
+        address creditor;
+        address srcVault;
+        address dstVault;
+        uint256 amount;
+        uint256 timestamp;
+    }
+}
+
+contract VaultQueue is ReentrancyGuardUpgradeable, Ownable {
+    mapping(address => IVaultQueue.Transfer[]) public qTransfer;
+    mapping(address => uint256) public totalAmount;
+    address[] public vaults;
+
+    function interVaultTransfer() external payable onlyOwner nonReentrant {
+        for (uint256 i = 0; i < vaults.length; i++) {
+            IRibbonVault(vaults[i]).completeWithdraw();
+        }
+
+        for (uint256 i = 0; i < vaults.length; i++) {
+            // Change to ERC20 instead of eth balance
+            uint256 balance = address(this).balance;
+
+            uint256 len = qTransfer[vaults[i]].length;
+            uint256 totalAmt = totalAmount[vaults[i]];
+            totalAmount[vaults[i]] = 0;
+
+            for (uint256 j = len; j > 0; j--) {
+                IVaultQueue.Transfer memory queue = pop(qTransfer[vaults[i]]);
+
+                uint256 portion =
+                    DSMath.wmul(balance, DSMath.wdiv(queue.amount, totalAmt));
+                require(portion > 0, "!portion");
+
+                IRibbonVault(queue.dstVault).depositFor{value: portion}(
+                    queue.creditor
+                );
+            }
+        }
+    }
+
+    function queueTransfer(
+        address srcVault,
+        address dstVault,
+        uint256 amount
+    ) external nonReentrant {
+        for (uint256 i = 0; i < vaults.length; i++) {
+            for (uint256 j = 0; j < qTransfer[vaults[i]].length; j++) {
```

Comment:
Worth refactoring this logic out to a view so the frontend can check too

Reply:
Done